src/phase0/state_transition/epoch_processing.rs

```diff
@@ -444,6 +530,433 @@ pub fn process_participation_record_updates<
+    get_attestation_component_deltas(state, &matching_target_attestations, context)
+}
+
+pub fn get_head_deltas<
+    const SLOTS_PER_HISTORICAL_ROOT: usize,
+    const HISTORICAL_ROOTS_LIMIT: usize,
+    const ETH1_DATA_VOTES_BOUND: usize,
+    const VALIDATOR_REGISTRY_LIMIT: usize,
+    const EPOCHS_PER_HISTORICAL_VECTOR: usize,
+    const EPOCHS_PER_SLASHINGS_VECTOR: usize,
+    const MAX_VALIDATORS_PER_COMMITTEE: usize,
+    const PENDING_ATTESTATIONS_BOUND: usize,
+>(
+    state: &BeaconState<
+        SLOTS_PER_HISTORICAL_ROOT,
+        HISTORICAL_ROOTS_LIMIT,
+        ETH1_DATA_VOTES_BOUND,
+        VALIDATOR_REGISTRY_LIMIT,
+        EPOCHS_PER_HISTORICAL_VECTOR,
+        EPOCHS_PER_SLASHINGS_VECTOR,
+        MAX_VALIDATORS_PER_COMMITTEE,
+        PENDING_ATTESTATIONS_BOUND,
+    >,
+    context: &Context,
+) -> Result<(Vec<Gwei>, Vec<Gwei>), Error> {
+    // Return attester micro-rewards/penalties for head-vote for each validator.
+    let previous_epoch = get_previous_epoch(state, context);
+    let matching_head_attestations =
+        get_matching_head_attestations(state, previous_epoch, context)?;
+    get_attestation_component_deltas(state, &matching_head_attestations, context)
+}
+
+pub fn get_inclusion_delay_deltas<
+    const SLOTS_PER_HISTORICAL_ROOT: usize,
+    const HISTORICAL_ROOTS_LIMIT: usize,
+    const ETH1_DATA_VOTES_BOUND: usize,
+    const VALIDATOR_REGISTRY_LIMIT: usize,
+    const EPOCHS_PER_HISTORICAL_VECTOR: usize,
+    const EPOCHS_PER_SLASHINGS_VECTOR: usize,
+    const MAX_VALIDATORS_PER_COMMITTEE: usize,
+    const PENDING_ATTESTATIONS_BOUND: usize,
+>(
+    state: &BeaconState<
+        SLOTS_PER_HISTORICAL_ROOT,
+        HISTORICAL_ROOTS_LIMIT,
+        ETH1_DATA_VOTES_BOUND,
+        VALIDATOR_REGISTRY_LIMIT,
+        EPOCHS_PER_HISTORICAL_VECTOR,
+        EPOCHS_PER_SLASHINGS_VECTOR,
+        MAX_VALIDATORS_PER_COMMITTEE,
+        PENDING_ATTESTATIONS_BOUND,
+    >,
+    context: &Context,
+) -> Result<(Vec<Gwei>, Vec<Gwei>), Error> {
+    // Return proposer and inclusion delay micro-rewards/penalties for each validator.
+    let previous_epoch = get_previous_epoch(state, context);
+    let validator_count = state.validators.len();
+    let mut rewards = vec![0; validator_count];
+    let mut matching_source_attestations = Vec::new();
+    for a in get_matching_source_attestations(state, previous_epoch, context)?.iter() {
+        matching_source_attestations.push(a);
+    }
+    for i in get_unslashed_attesting_indices(state, &matching_source_attestations, context)? {
+        let mut attestations = Vec::new();
+        for a in &matching_source_attestations {
+            if get_attesting_indices(state, &a.data, &a.aggregation_bits, context)?.contains(&i) {
+                attestations.push(*a)
+            }
+        }
+        let attestation = attestations
+            .iter()
+            .min_by(|&a, &b| a.inclusion_delay.cmp(&b.inclusion_delay))
+            .ok_or(Error::InvalidAttestation)?;
```

Comment:
I wasn't sure what's ideal for error handling an `Option`, so I opted for `ok_or` instead of doing some `match`. I also wasn't sure exactly what to use -- sort of inserted `InvalidAttestation` as it was already an existing type, but I'm not sure if it's what it's intended for.

Reply:
so we can consider what kind of error this is and use that to inform what the `Option` here really *means*

and we can see that `min_by` returns `None` only when `attestations` would be empty, and we know that `attestations` will never be empty as our "source" material is ultimately from the `matching_source_attestations`

so in this case, we can actually just `unwrap` the option -- and in fact i'd suggest using `expect` with a message like "at least one attestation in collection" (as it provides helpful context if something does blow up unexpectedly)